csrc/scheduler/hopper_multi_matmul.cpp

```diff
@@ -990,79 +990,83 @@ void HopperMultipleMatmulScheduler::scheduleOutputTensor(TensorView* c) {
-    d->axis(-1)->parallelize(ParallelType::Vectorize);
-  }
-  return;
-  std::vector<TensorView*> output_tvs;
-  for (Val* v : fusion_->outputs()) {
-    if (auto tv = dynamic_cast<TensorView*>(v)) {
-      output_tvs.push_back(tv);
-    }
-  }
-  if (params_->use_smem_epilogue) {
-    blockTileTensors(output_tvs);
-    for (auto [dc, d] : cached_outputs_) {
-      // Schedule output tensor differently for better global memory access
-      // pattern.
-      scheduleOutputTensor(d);
+  if (!params_->use_smem_epilogue) {
+    for (auto& [dc, d] : cached_outputs_) {
+      blockTileTensors({dc});
+      blockTileTensors({d});
+      for (auto tv : {dc, d}) {
+        // [..., Mo, No, Mi, Ni]
+        tv->split(-2, getM(params_->mma_macro));
+        tv->split(-1, getN(params_->mma_macro));
+        // [..., Mo, No, Mio, Mii, Nio, Nii]
+        // -> [..., Mo, No, Mio, Nio, Mii, Nii]
+        tv->reorder({{-3, -2}});
+        tv->merge(-4);
+        auto s = mma_utils::MmaSwizzler::scheduleMmaOutputAllocation(
+            tv->getLoopDomain());
+        tv->setLoopDomain(s.as<IterDomain*>());
+        tv->axis(-5)->parallelize(ParallelType::TIDy);
+      }
+      parallelizeBlocks({dc});
+      parallelizeBlocks({d});
       d->axis(-1)->parallelize(ParallelType::Vectorize);
-
-      // Propagate output tensor transformations back to smem_epilogue
-      scheduler_utils::BoundedDirectionalTransformPropagator::backward(
-          d, -1, smem_epilogues_);
     }
   } else {
-    for (TensorView* mma_result : mma_results_) {
-      scheduler_utils::BoundedDirectionalTransformPropagator::forward(
-          mma_result,
-          -1,
-          output_tvs,
-          scheduler_utils::BoundedDirectionalTransformPropagator::Options()
-              .propagateParallelType()
-              .propagateToBoundary());
-    }
-    for (auto [dc, d] : cached_outputs_) {
-      // We might propagate an inner dimension that is not compatible with the
-      // output or bias-like inputs. In those cases, we will further split
-      // this dimension with an outer unrolled loop to achieve the proper
-      // vectorization as specified by params.supported_vec_size.epilogue.
-      NVF_ERROR(d->axis(-1)->extent()->isConst());
-      int64_t d_extent = d->axis(-1)->extent()->value().as<int64_t>();
-      if (d_extent > params_->supported_vec_size.epilogue) {
-        // Should always be a divisible split
-        NVF_ERROR(d_extent % params_->supported_vec_size.epilogue == 0);
-        d->split(
-            -1, params_->supported_vec_size.epilogue, /*inner_split=*/true);
-        d->axis(-2)->parallelize(ParallelType::Unroll);
+    // TODO: Use stmatrix to load from registers to shared memory
+    constexpr int64_t stmatrix_tile_m = 16;
+    constexpr int64_t stmatrix_tile_n = 16;
+    constexpr int64_t tma_m = 16;
+    constexpr int64_t tma_n = 64;
+
+    // Manually schedule register cache and output TensorView
+    for (auto& [dc, d] : cached_outputs_) {
```

Comment:
FYI I am thinking of removing `cached_outputs_` because it gets out of sync easily. For example when we translate matmul patterns, sometimes we need to replace the pattern output, and sometimes that pattern output is a `dc`. In thayose cases we wind up needing to figure out the replacement and update `cached_outputs_`. But we don't really need that anyway since `fusion_->outputs()` gives us all the `d` tensors and `d->definition()->input(0)->as<TensorView>()` or `ir_utils::producerTensorsOf(d).first()` gives us `dc`.

Reply:
https://github.com/NVIDIA/Fuser/pull/3478